KEYCLOAK-15770 Skip creating session for docker protocol authentication

mposolda opened the pull request:
@stianst @hmlnarik @pedroigor The PR tries to skip creating of user/client sessions after successful docker protocol authentication as it seems this is not needed. Hopefully it can be re-used for skip creating of sessions in other use-cases (Especially OAuth2 Client Credentials Grant. But maybe even some more like SAML ECP?).

I've tried to use the approach that userSession/clientSession are still created as objects, but those are just "transient" sessions, which are not saved to infinispan cache at all. They are valid just for the scope of current request (current KeycloakSession lifecycle). This makes it easier to work with protocolMappers SPI and other parts of the code, which rely on the fact that userSession/clientSession objects exist and avoid some NPE and other similar issues...

WDYT? Do you please have a chance to review?

mposolda committed:
KEYCLOAK-15770 Skip creating session for docker protocol authentication

pedroigor reviewed (commented):
Interesting this concept of transient user sessions. Why not have that on the session contract then and not specific to the login protocol in use ?

mposolda commented:
> Interesting this concept of transient user sessions. Why not have that on the session contract then and not specific to the login protocol in use ?

@pedroigor Thanks. I am not sure what exactly you mean, but it is highly possible that just the "protocol" may not be enough information to detect if session can be transient. Maybe some more context data might be needed. For example for Docker protocol, there is a chance that we must be able to create the session in some cases (This is still to be discussed and investigated regarding docker protocol). Also for example SAML or OIDC may usually require session, but maybe for some cases it doesn't require sessions? (For example OIDC client credentials may not need session as we discussed in other PR).

hmlnarik commented:
> > Interesting this concept of transient user sessions. Why not have that on the session contract then and not specific to the login protocol in use ?
> 
> @pedroigor Thanks. I am not sure what exactly you mean, but it is highly possible that just the "protocol" may not be enough information to detect if session can be transient. Maybe some more context data might be needed. For example for Docker protocol, there is a chance that we must be able to create the session in some cases (This is still to be discussed and investigated regarding docker protocol). Also for example SAML or OIDC may usually require session, but maybe for some cases it doesn't require sessions? (For example OIDC client credentials may not need session as we discussed in other PR).

I like the addition of the transient session, thank you!
I second @pedroigor's idea. Could the transient flag could actually be based on e.g. some note in authenticationSession rather than be hardcoded in the login protocol?
One nit: perhaps an `Enum` would server better for readabililty than a `boolean` constant in the `createUserSession` method?

mposolda commented:
> I like the addition of the transient session, thank you!
> I second @pedroigor's idea. Could the transient flag could actually be based on e.g. some note in authenticationSession rather than be hardcoded in the login protocol?
> One nit: perhaps an `Enum` would server better for readabililty than a `boolean` constant in the `createUserSession` method?

Ah, ok. Yes, I can change it to have a note on the authSession, which will drive if userSession can be transient or not. One side-effect is, that individual authenticators will have a chance to set this note and change the flag, which is probably an advantage and more flexibility. Also I will change the `Enum` in `createUserSession` method. Thanks for the feedback!

mposolda commented:
@hmlnarik @pedroigor I did more updates to the PR per your feedback. Summary:
- Added enum UserSessionModel.SessionPersistenceState instead of true/false as argument to "createUserSession" method as was before
- Used AuthenticationSession note to track if userSession should be persistent or transient. Removed changes from LoginProtocol and related class
- DockerProtocol is still the only built-in login protocol, which uses transient sessions. All OIDC and SAML flows still use persistent session as was before
- Added TransientSessionTest (as the Docker protocol automated test does not work)

pedroigor commented on testsuite/integration-arquillian/servers/auth-server/services/testsuite-providers/src/main/java/org/keycloak/testsuite/forms/SetClientNoteAuthenticator.java:
```diff
@@ -0,0 +1,150 @@
+
+    @Override
+    public void authenticate(AuthenticationFlowContext context) {
+        MultivaluedMap<String, String> inputData = context.getHttpRequest().getDecodedFormParameters();
```
Do we really want to push whatever comes from the request to the session notes? Does not look safe for me ...

I think that ideally, we should have some switch to enable/disable this on a per-flow basis or have a `Transient Session Authenticator` that always set this note to indicate whether or not sessions for that flow should be transient.

pedroigor replied:
Or maybe this is something that should be hardcoded for now into Docker protocol which is the initial problem we want to solve.

mposolda replied:
@pedroigor Sure, this is not safe. This is testsuite-only authenticator, which is not intended to be ever used in production. I've used authenticator just to be able to test this for the OAuth2 DirectGrant, as the Docker protocol automated tests don't work.

As you pointed, in case of the docker protocol, it is currently hardcoded - the note is set in the DockerEndpoint class. Other protocols (OIDC, SAML) currently don't have support for transient session in this PR - with the only exception of the testsuite, which have this SetClientNoteAuthenticator .

Do you think that we need authenticator to be added to the main codebase to be able to have transient sessions per flow? I can think that OAuth2 Resource Owner Password Credentials Grant (Direct Grant) in some cases doesn't need user session at all. AFAIK Sometimes people use DirectGrant just for the single authentication to some CLI app and never need to refresh or introspect tokens etc. Not sure if this is useful also for some other protocols - like for example SAML ECP? WDYT?

pedroigor replied:
Ohh, sorry. Didn't notice it was related to the test suite.

But yeah, I'm OK with having something in the main codebase to achieve that. Not sure if the authenticator approach is the best path though ... At the same time, I'm not sure if adding a config option to authentication flows is something we want to do.

SAML ECP is a good example, yeah.

Btw, transient sessions should not work if later you try to introspect the token using the token introspection endpoint, right? I mean, people should be aware that the lifetime of tokens issued using a transient session is bound to their time-related claims and can/should only be validated locally by the resource server. And no support for revocation but wait until the time-related claims become invalid.

hmlnarik replied:
> Btw, transient sessions should not work if later you try to introspect the token using the token introspection endpoint, right?

That's a good point, there should be a docs update in https://www.keycloak.org/docs/latest/server_admin/#_docker

The PR looks good to me as is once the docs are in place. My proposal would be to merge this PR as is once the docs are ready, and include no authenticator or other means to set the client note in the main code base now. It is straightforward to mark the session to be transient, and the best means to do so may depend on the particular use case. @pedroigor WDYT?

mposolda replied:
@hmlnarik @pedroigor Well, I am starting to think that Authenticator approach may be nice and will have some advantages :) Administrator can decide if he wants to create transient session by simply adding the authenticator to the authentication flow. May be useful for DirectGrant or SAML ECP. For Docker protocol, we can have this authenticator in the flow by default, but admin will have a chance to remove it if he wants and have sessions persistent for docker as well for some reason.

Authenticatior will have some additional flexibility that administrator may want to persist sessions just under specified condition (hence the authenticator can be added to some conditional subflow and be triggered just in case of the condition passed)... This is something, which won't be easily available if we have this as hardcoded option of the authentication flow.

On the other hand, this won't be classic "Authenticator", but rather some kind of interceptor in the authentication flow. But our Authentication SPI don't have this concept, so just use it as implementation of "Authenticator" interface can be possible.

So my ideal way would be to change this PR to have the Authenticator approach, but if you prefer, I can keep it as is and just update the documentation. WDYT?

I agree that docs will need to be updated. As pointed, transient session would mean that introspection, "refresh token" or "Revoke session" won't be possible. I will update the docs once we refine whether to go with authenticator approach or keep the PR as is.

hmlnarik replied:
The authenticator approach it is certainly interesting and possible way to deal with it, yet I am not convinced yet it's the right way. For example, there is no separate SAML / ECP flow so it would not be possible to use an authenticator in this case.

At this moment, I'd proceed with this PR as is, and only follow up with a common approach once we have it.

pedroigor replied:
I agree. I'm still not 100% sure about the Authenticator but is what we have ... @mposolda I agree this looks like a new concept (or, potentially, a new configuration option for Authentication Flows).

In any case, LGTM. Perhaps we should make this somehow tech-preview and advertize that it may change in the future.

mposolda replied:
@hmlnarik @pedroigor Ok, Thanks. Let's wait with the authenticator or something else to later if needed. The docs PR is here https://github.com/keycloak/keycloak-documentation/pull/1045 .

I did not know SAML ECP profile does not have separate flow, Thanks for clarifying. I somehow assumed that yes based on the fact that Direct grant (OAuth2 Resource Owner PAssword Credentials) has it and I though this is somehow similar.
